### PR TITLE
amp-ad-exit click delay interval start based on preformance timing event

### PR DESCRIPTION
**Pull request by keithwrightbos**
Currently click delay interval starts at extension initialization which can depend on cache hit rate and other factors.  Allow for basing on performance timing events (e.g. navigationStart) via startTimingEvent in the config.  If browser does not support performance timing or event given does not exist in timing object, click protection will not be applied.

### CODE REVIEW

File: extensions/amp-ad-exit/0.1/filters/click-delay.js [hunk 1]
```diff
@@ -21,32 +21,34 @@ export class ClickDelayFilter extends Filter {
...
-  constructor(name, spec) {
+  constructor(name, spec, win) {
     super(name);
-    user().assert(isValidClickDelaySpec(spec), 'Invalid ClickDelay spec');
+    user().assert(spec.type == FilterType.CLICK_DELAY &&
+      typeof spec.delay == 'number' && spec.delay > 0,
+    'Invalid ClickDelay spec');
+    user().assert(!spec.startTimingEvent ||
+      (win['performance'] && win['performance']['timing']),
+    'Browser does not support performance timing');
```
Comment (clawr) [inline extensions/amp-ad-exit/0.1/filters/click-delay.js]: Should this be a fatal problem? Should it be the responsibility of the creative author to anticipate which browsers their creative will serve to and know which web platform features are available? I'd rather treat this property as a best-effort feature upgrade.

Comment (jeffkaufman) [reply extensions/amp-ad-exit/0.1/filters/click-delay.js]: If performance timing isn't available, we should probably fall back to 1s from when we first got to run our code, no?

Comment (keithwrightbos) [reply extensions/amp-ad-exit/0.1/filters/click-delay.js]: Done - falls back to using time interval when extension executed

File: extensions/amp-ad-exit/0.1/test/filters/test-click-delay.js [hunk 2]
```diff
@@ -24,41 +24,54 @@ describe('click-delay', () => {
...
+  let sandbox;
+  beforeEach(() => sandbox = sinon.sandbox.create());
+  afterEach(() => sandbox.restore());
 
-  describe('should fail on invalid configs', () => {
+  it('should use performance timing', () => {
+    const win = {performance: {timing: {'navigationStart': 456}}};
+    sandbox.stub(Date, 'now').returns(123);
+    expect(new ClickDelayFilter('foo', DEFAULT_CONFIG, win).intervalStart)
+        .to.equal(123);
```
Comment (clawr) [inline extensions/amp-ad-exit/0.1/test/filters/test-click-delay.js]: shouldn't this be 456?

File: extensions/amp-ad-exit/0.1/test/filters/test-click-delay.js [hunk 3]
```diff
@@ -24,41 +24,54 @@ describe('click-delay', () => {
...
-      nowStub.onFirstCall().returns(2);
-      nowStub.onSecondCall().returns(125);
       const filter = new ClickDelayFilter(
           'foo', DEFAULT_CONFIG, {performance: {timing: {navigationStart: 1}}});
+      const nowStub = sandbox.stub(Date, 'now');
+      nowStub.onFirstCall().returns(1);
+      nowStub.onSecondCall().returns(125);
       expect(filter.filter()).to.be.false;
       expect(filter.filter()).to.be.true;
     });
```
Comment (clawr) [inline extensions/amp-ad-exit/0.1/test/filters/test-click-delay.js]: could you add another test where navigationStart = 1, Date.now() always returns the same value >= 1001, and check that filter() is true? To test that the Filter properly compares now() to the specified timing event.

File: extensions/amp-ad-exit/0.1/test/filters/test-click-delay.js (hunk 2 above)
Comment (keithwrightbos) [reply extensions/amp-ad-exit/0.1/test/filters/test-click-delay.js]: good catch... had a typo!

**Comment by clawr** [comment]
LGTM

**Pull request merged**